src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/Utils/BinaryReaderExtensions.cs

```diff
@@ -70,36 +74,67 @@ internal static object ReadPrimitiveValue(this BinaryReader reader, PrimitiveTyp
-
-        // Use DateTime's public constructor to validate the input, but we
-        // can't return that result as it strips off the kind. To address
-        // that, store the value directly into a DateTime via an unsafe cast.
-        // See BinaryFormatterWriter.WriteDateTime for details.
+        ulong ticks = dateData & 0x3FFFFFFF_FFFFFFFFUL;
+        DateTimeKind kind = (DateTimeKind)(dateData >> 62);
 
         try
         {
-            const long TicksMask = 0x3FFFFFFFFFFFFFFF;
-            _ = new DateTime(data & TicksMask);
+            return ((uint)kind <= (uint)DateTimeKind.Local) ? new DateTime((long)ticks, kind) : CreateFromAmbiguousDst(ticks);
         }
         catch (ArgumentException ex)
         {
-            // Bad data
             throw new SerializationException(ex.Message, ex);
         }
 
-        return Unsafe.As<long, DateTime>(ref data);
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static DateTime CreateFromAmbiguousDst(ulong ticks)
+        {
+            // There's no public API to create a DateTime from an ambiguous DST, and we
+            // can't use private reflection to access undocumented .NET Framework APIs.
+            // However, the ISerializable pattern *is* a documented protocol, so we can
+            // use DateTime's serialization ctor to create a zero-tick "ambiguous" instance,
+            // then keep reusing it as the base to which we can add our tick offsets.
+
+            if (s_baseAmbiguousDstDateTime is not DateTime baseDateTime)
+            {
+#pragma warning disable SYSLIB0050 // Type or member is obsolete
+                SerializationInfo si = new(typeof(DateTime), new FormatterConverter());
+                si.AddValue("ticks", 0L); // legacy value (serialized as long) - specify both just to be safe
+                si.AddValue("dateData", 0xC0000000_00000000UL); // new value (serialized as ulong)
+
+#if NET
+                baseDateTime = CallPrivateSerializationConstructor(si, new StreamingContext(StreamingContextStates.All));
+#else
+                ConstructorInfo ci = typeof(DateTime).GetConstructor(
+                    BindingFlags.Instance | BindingFlags.NonPublic,
+                    binder: null,
+                    new Type[] { typeof(SerializationInfo), typeof(StreamingContext) },
+                    modifiers: null);
+
+                baseDateTime = (DateTime)ci.Invoke(new object[] { si, new StreamingContext(StreamingContextStates.All) });
+#endif
+
+#pragma warning restore SYSLIB0050 // Type or member is obsolete
+                Volatile.Write(ref s_baseAmbiguousDstDateTime, baseDateTime); // it's ok if two threads race here
+            }
+
+            return baseDateTime.AddTicks((long)ticks);
+        }
+
+#if NET
+        [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
+        extern static DateTime CallPrivateSerializationConstructor(SerializationInfo si, StreamingContext ct);
+#endif
```

Comment:
Using a private constructor like this from a separate package is considered safe / supported?

Reply:
I agree. `System.Formats.Nrbf` should not depend on any of the built-in legacy infrastructure for binary serialization.

Reply:
> AFAIK we have no plans to remove these ctors.

We've already made some of them throw PlatformNotSupportedException, e.g.
https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs#L175-L176
and entirely removed them from others, e.g.
https://github.com/dotnet/runtime/blob/0f05719bd26563383d826f3dcbde53879baad58f/src/libraries/System.Private.CoreLib/src/System/OperatingSystem.cs#L12

Reply:
Should we have a public API that allows us to create the specific data time value?

Reply:
> Should we have a public API that allows us to create the specific data time value?

Yes, but it would not solve the problem as this package needs to support older monikers, including `netstandard2.0`

Reply:
I think both Unsafe.As and reflection are ok for existing targets. The existing targets are set in stone and we can make assumptions about them.

Unsafe.As or reflection are less than ideal for future. They limit the changes we can do in future.